lisa/sut_orchestrator/libvirt/platform.py

```diff
@@ -1341,3 +1388,316 @@ def _capture_libvirt_logs(self) -> None:
+                        results.append(combo)
+                        break
+                if results:
+                    break
+
+        if not results:
+            raise SkippedException(
+                f"Pool {pool_type} running out of devices: {pool}, "
+                "No IOMMU Group has sufficient count of devices, "
+                f"Refer: {pool}"
+            )
+
+        devices: List[DeviceAddressSchema] = []
+        selected_pools = results[0]
+        for iommu_grp in selected_pools:
+            devices += pool.pop(iommu_grp)
+        self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+        return devices
+
+    def _put_devices_into_pool(
+        self,
+        node_context: NodeContext,
+    ) -> None:
+        device_context = node_context.device_passthrough_context
+        for context in device_context:
+            pool_type = context.pool_type
+            devices_list = context.device_list
+            pool = self.AVAILABLE_HOST_DEVICES.get(pool_type, {})
+            for device in devices_list:
+                iommu_grp = self._get_device_iommu_group(device)
+                pool_devices = pool.get(iommu_grp, [])
+                pool_devices.append(device)
+                pool[iommu_grp] = pool_devices
+            self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+
+    def _verify_device_passthrough_post_boot(
+        self,
+        node_context: NodeContext,
+    ) -> None:
+        device_context = node_context.device_passthrough_context
+        for context in device_context:
+            devices = context.device_list
+            for device in devices:
+                err = f"Kernel driver is not vfio-pci for device: {device}"
+                pool_type = context.pool_type.value
+                if context.managed == "yes" and "pci" in pool_type:
+                    is_vfio_pci = self._is_driver_vfio_pci(device)
+                    assert is_vfio_pci, err
+
+    def _check_passthrough_support(self, node: Node) -> None:
+        ls = node.tools[Ls]
+        path = "/dev/vfio/vfio"
+        err = "Host does not support IOMMU"
+        if not ls.path_exists(path=path, sudo=True):
+            raise LisaException(f"{err} : {path} does not exist")
+
+        path = "/sys/kernel/iommu_groups/"
+        if len(ls.list(path=path, sudo=True)) == 0:
+            raise LisaException(f"{err} : {path} does not have any entry")
+
+    def _is_driver_vfio_pci(
+        self,
+        device_addr: DeviceAddressSchema,
+    ) -> bool:
+        lspci = self.host_node.tools[Lspci]
+        device_addr_str = self._get_pci_address_str(device_addr)
+        kernel_module = lspci.get_used_module(device_addr_str)
+        return kernel_module == "vfio-pci"
+
+    def _set_device_passthrough_node_context(
+        self,
+        node_context: NodeContext,
+        node_runbook: BaseLibvirtNodeSchema,
+    ) -> None:
+        if node_runbook.device_passthrough:
+            node_context.is_device_passthrough_set = True
```

Comment:
Is the `is_device_passthrough_set` variable needed? Could you do the same check by looking at if the `device_passthrough_context` list is empty or not?

Reply:
@squirrelsc @cwize1 Used length of the list now.